### PR TITLE
feat(*): add tencent cloud provider and clb plugin

### PR DESCRIPTION
**Pull request by imroc**
## About me

Thanks a lot for this awesome project! I'm [roc](https://imroc.cc), an engineer from the Tencent Kubernetes Engine(TKE) team.

## Background

TKE has [tke-extend-network-controller](https://github.com/tkestack/tke-extend-network-controller), which supports network access for room-based games and exposes game rooms by binding game server pod to CLB(Cloud Load Balancer) listener exclusively(1 CLB listener, 1 POD ip:port).

`tke-extend-network-controller` does not rely on OpenKruiseGame(OKG) or Agones, But OKG users prefer to define the game room network in the OKG way, this PR add the tencentcloud provider to OKG and add the clb plugin, which based on `tke-extend-network-controller`'s `DedicatedCLBListener`(Atomic capability to bind a pod ip:port to a clb's port).

## Why not based on LoadBalaner service ?

Others OKG lb plugins (e.g. AlibabaCloud-SLB, Volcengine-CLB) are based on LoadBalancer service, which use annotation to reuse existing LoadBalaner, and define lb port on the LoadBalancer service.

But for TKE, a lot of LoadBalancer service reuse the same LoadBalancer is not recommended and disabled by default, this is mainly because of the service-controller's reconciliation performance issues for this scenario, which may cause a long delay in the binding or unbinding operation, and it is difficult to optimize due to historical and architectural reasons. This is also why `tke-extend-network-controller` introduce `DedicatedCLBListener` rather than based on LoadBalancer service to reuse clb to bind pods.

## How to use

This plugin is for TKE Users, User need install [tke-extend-network-controller](https://github.com/tkestack/tke-extend-network-controller) first (can be installed through TKE application marketplace), then make sure `tencentcloud` provider is enabled:

```toml
[tencentcloud]
enable = true
[tencentcloud.clb]
max_port = 1100
min_port = 1000
```

And define `TencentCloud-CLB` network for `GameServerSet`:

```yaml
apiVersion: game.kruise.io/v1alpha1
kind: GameServerSet
metadata:
  name: clb-nginx
  namespace: default
spec:
  replicas: 2
  updateStrategy:
    rollingUpdate:
      podUpdatePolicy: InPlaceIfPossible
  network:
    networkType: TencentCloud-CLB
    networkConf:
      - name: ClbIds
        value: "lb-3ip9k5kr,lb-4ia8k0yh"
      - name: PortProtocols
        value: "80/TCP,7777/UDP"
  gameServerTemplate:
    spec:
      containers:
        - image: nginx
          name: nginx
```

You could found auto created dedicatedclblisteners:

```bash
$ kubectl get dedicatedclblisteners.networking.cloud.tencent.com
NAME                LBID          LBPORT   POD           STATE
clb-nginx-0-7ndzj   lb-3ip9k5kr   1003     clb-nginx-0   Bound
clb-nginx-0-gqmbm   lb-3ip9k5kr   1002     clb-nginx-0   Bound
clb-nginx-1-pdckl   lb-3ip9k5kr   1001     clb-nginx-1   Bound
clb-nginx-1-zk9st   lb-3ip9k5kr   1000     clb-nginx-1   Bound
```

And you can also found clb's ip:port(externalAddresses) for gameserver pod on GameServer's networkStatus:

```yaml
  networkStatus:
    createTime: "2024-10-28T03:16:20Z"
    currentNetworkState: Ready
    desiredNetworkState: Ready
    externalAddresses:
    - ip: 139.155.64.52
      ports:
      - name: "80"
        port: 1002
        protocol: TCP
    - ip: 139.155.64.52
      ports:
      - name: "7777"
        port: 1003
        protocol: UDP
    internalAddresses:
    - ip: 172.16.7.106
      ports:
      - name: "80"
        port: 80
        protocol: TCP
    - ip: 172.16.7.106
      ports:
      - name: "7777"
        port: 7777
        protocol: UDP
    lastTransitionTime: "2024-10-28T03:16:20Z"
    networkType: TencentCloud-CLB
```

Check more info in `user_manuals/network.md`.

## What's Next

* Add RBAC permission for `networking.cloud.tencent.com` apiGroup in kruise-game chart's [rbac_role.yaml](https://github.com/openkruise/charts/blob/master/versions/kruise-game/0.9/templates/rbac_role.yaml#L80) after this PR is merged.
* Add doc for [openkruise.io](https://github.com/openkruise/openkruise.io) after this PR is merged.

### CODE REVIEW

File: docs/en/user_manuals/network.md [hunk 1]
```diff
@@ -969,4 +969,107 @@ nlb-26jbknebrjlejt5abu      192.168.0.8:80,192.168.0.82:80,192.168.63.228:80
...
+
+#### Plugin description
+
+
+- TencentCloud-CLB enables game servers to be accessed from the Internet by using Cloud Load Balancer (CLB) of Tencent Cloud. CLB is a type of Server Load Balancer (CLB). TencentCloud-CLB uses different ports for different game servers. The CLB instance only forwards traffic, but does not implement load balancing.
+The [tke-extend-network-controller](https://github.com/tkestack/tke-extend-network-controller) network plugin needs to be installed (can be installed through the TKE application market).
+- This network plugin does not support network isolation.
+
+#### Network parameters
+
```
Comment (chrisliu1995) [inline docs/en/user_manuals/network.md:994]: lack the name of ClbIds

File: cloudprovider/manager/provider_manager.go [hunk 2]
```diff
@@ -138,5 +140,15 @@ func NewProviderManager() (*ProviderManager, error) {
 		}
 	}
 
+	if configs.TencentCloudOptions.Valid() && configs.TencentCloudOptions.Enabled() {
+		// build and register amazon web services provider
```
[the remark this answers is not in the document]
Comment (imroc) [reply cloudprovider/manager/provider_manager.go]: nice catch, fixed

File: docs/en/user_manuals/network.md (hunk 1 above)
Comment (imroc) [reply docs/en/user_manuals/network.md:994]: fixed

File: apis/v1alpha1/gameserverset_types.go [hunk 3]
```diff
@@ -30,6 +30,7 @@ import (
 
 const (
 	GameServerOwnerGssKey      = "game.kruise.io/owner-gss"
+	OwnerPodKey                = "game.kruise.io/owner-pod"
```
[the remark this answers is not in the document]
Comment (imroc) [reply apis/v1alpha1/gameserverset_types.go:33]: Auto-created `DedicatedCLBListener` is related a single pod, one pod with multiple ports have multiple `DedicatedCLBListener`,  each `DedicatedCLBListener` has a label `game.kruise.io/owner-pod`, so we can use label selector to find pod related `DedicatedCLBListener` in `OnPodUpdated` callback to reconcile pod related `DedicatedCLBListener`.

**Comment by imroc** [comment]
e2e failure is not related, re-run could succeed (new commit only have doc change)

File: apis/v1alpha1/gameserverset_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisliu1995) [reply apis/v1alpha1/gameserverset_types.go:33]: I suggest to put that key into package of cloud provider. Plus, you can also use the label "statefulset.kubernetes.io/pod-name" to select pod by pod-name. Every pod has that label.

**Comment by chrisliu1995** [comment]
> e2e failure is not related, re-run could succeed (new commit only have doc change)

File: apis/v1alpha1/gameserverset_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (imroc) [reply apis/v1alpha1/gameserverset_types.go]: OK, moved into package of cloud provider. This is used to select `DedicatedCLBListener` by pod-name, not to select pod.

**Pull request merged**